lib/scholar/linear/logistic_regression.ex

```diff
@@ -85,127 +83,99 @@ defmodule Scholar.Linear.LogisticRegression do
-    x_t = Nx.transpose(x)
+    y = Scholar.Preprocessing.one_hot_encode(y, num_classes: num_classes)
 
-    {_, _, _, _, _, _, _, final_coeff} =
-      while {iter = 0, x, learning_rate, n, iterations, one_hot, x_t,
-             coeff = Nx.broadcast(Nx.tensor(0, type: Nx.type(x)), {n, num_classes})},
+    {_, _, _, _, _, _, final_coeff, final_bias} =
+      while {iter = 0, x, learning_rate, n, iterations, y,
+             coeff =
+               Nx.broadcast(
+                 Nx.tensor(1.0, type: Nx.Type.to_floating(Nx.type(x))),
+                 {n, num_classes}
+               ),
+             bias =
+               Nx.broadcast(Nx.tensor(0, type: Nx.Type.to_floating(Nx.type(x))), {num_classes})},
             iter < iterations do
-        coeff = update_coefficients_multinomial(x, x_t, one_hot, coeff, learning_rate)
-        {iter + 1, x, learning_rate, n, iterations, one_hot, x_t, coeff}
+        {coeff_grad, bias_grad} = grad_loss(coeff, bias, x, y)
+        coeff = coeff - learning_rate * coeff_grad
+        bias = bias - learning_rate * bias_grad
+        {iter + 1, x, learning_rate, n, iterations, y, coeff, bias}
       end
 
     %__MODULE__{
-      coefficients: final_coeff,
-      bias: Nx.tensor(0, type: Nx.type(x)),
-      mode: :multinomial
+      coefficients: Nx.transpose(final_coeff),
+      bias: final_bias
     }
   end
 
-  # Normalized softmax
-
-  defnp softmax(t) do
-    normalized_exp = (t - Nx.reduce_max(t, axes: [0], keep_axes: true)) |> Nx.exp()
-    normalized_exp / Nx.sum(normalized_exp, axes: [0])
+  defnp cross_entropy(log_probs, targets) do
+    target_class = Nx.argmax(targets, axis: 1)
+    nll = Nx.take_along_axis(log_probs, Nx.new_axis(target_class, 1), axis: 1)
+    -Nx.mean(nll)
   end
 
-  # Gradient descent for binary regression
-
-  defnp update_coefficients(x, x_t, y_t, {coeff, bias}, learning_rate) do
-    {m, _n} = x.shape
-
-    logit = 1 / (Nx.exp(-(Nx.dot(x, coeff) + bias)) + 1)
-
-    diff = Nx.reshape(logit - y_t, {m})
-
-    coeff_diff = Nx.dot(x_t, diff) / m
-
-    bias_diff = Nx.sum(diff) / m
-
-    new_coeff = coeff - coeff_diff * learning_rate
-    new_bias = bias - bias_diff * learning_rate
-
-    {new_coeff, new_bias}
+  defnp loss_fn(coeff, bias, xs, ys) do
+    probs = Axon.Activations.log_softmax(Nx.dot(xs, coeff) + bias)
+    cross_entropy(probs, ys)
   end
 
-  # Gradient descent for multinomial regression
-
-  defnp update_coefficients_multinomial(x, x_t, one_hot, coeff, learning_rate) do
-    {m, _n} = x.shape
-
-    dot_prod = Nx.dot(x, coeff)
-
-    prob = softmax(dot_prod)
+  defnp grad_loss(coeff, bias, xs, ys) do
+    grad({coeff, bias}, fn {coeff, bias} -> loss_fn(coeff, bias, xs, ys) end)
+  end
 
-    diff = Nx.dot(x_t, prob - one_hot) / m
+  # Normalized softmax
 
-    coeff - diff * learning_rate
+  defn softmax(t) do
+    max = Nx.Defn.Kernel.stop_grad(Nx.reduce_max(t, axes: [-1], keep_axes: true))
```

Comment:
everything in Nx.Defn.Kernel is imported inside defn.

```suggestion
    max = stop_grad(Nx.reduce_max(t, axes: [-1], keep_axes: true))
```